server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -484,6 +484,23 @@ private static NodeMetadata loadNodeMetadata(Settings settings, Logger logger,
         return metadata;
     }
 
+    /**
+     * loads existing node metadata from disk without attempting to upgrade to current version
+     */
+    public NodeMetadata loadLastKnownMetadata() {
```

Comment:
To me this method looks brittle because the updated version will be written back in https://github.com/elastic/elasticsearch/blob/f2ea4253f237d4bd6a6cb4c6f3f4406059a64cec/server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java#L90 , so this method can actually return either version, depending when it is invoked. It works right now because it is called from the `Node` constructor, but it could conceivably be moved in `Node#start` and there it might race with the write.

I don't have any better ideas though. I will defer to @DaveCTurner . Maybe it is reasonable to begin storing the version the node upgraded from, maybe it is useful in other places as well.

Reply:
++ yes I'd rather we captured the original node version in `loadNodeMetadata` rather than reading the metadata twice like this.

Reply:
The nodemetadata lifecycle was considered when implementing this. I've chatted with David and he was also of the view that it would be preferable to store this in the NodeMetadata rather than double reading it but was _ok_ too. Since it seems to be piling up, I have nothing against changing the approach. I'll wait for David to weigh in too so I make changes in bulk

Reply:
Sounds good, will amend